File: api/src/main/java/me/kvdpxne/dtm/team/LocalTeam.java
package me.kvdpxne.dtm.team;

import java.util.Collection;
import me.kvdpxne.dtm.game.LocalGame;
import me.kvdpxne.dtm.team.teammate.Teammate;
import me.kvdpxne.dtm.translation.Communicable;
import me.kvdpxne.dtm.user.LocalUser;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;

/**
 * @since 0.1.0
 */
public interface LocalTeam
  extends
  Communicable,
  Team {

  /**
   * @since 0.1.0
   */
  @NotNull
  LocalGame getLocalGame();

  /**
   * @since 0.1.0
   */
  @NotNull
  Collection<Teammate> getTeammates();

  /**
   * @since 0.1.0
   */
  @Range(from = 0, to = Integer.MAX_VALUE)
  int getHealth();

  /**
   * @since 0.1.0
   */
  @Range(from = 0, to = Integer.MAX_VALUE)
  int getSize();

  /**
   * @since 0.1.0
   */
  boolean hasTeammate(
    final @NotNull LocalUser user
  );

  /**
   * @since 0.1.0
   */
  boolean hasTeammate(
    final @NotNull Teammate teammate
  );

  /**
   * @since 0.1.0
   */
  @Nullable
  Teammate getTeammate(
    final @NotNull LocalUser user
  );

  /**
   * @since 0.1.0
   */
  boolean addTeammate(
    final @NotNull Teammate teammate
  );

  /**
   * @since 0.1.0
   */
  boolean removeTeammate(
    final @NotNull Teammate teammate
  );

  /**
   * @since 0.1.0
   */
  boolean removeTeammate(
    final @NotNull LocalUser localUser
  );

  /**
   * @since 0.1.0
   */
  @Range(from = 0, to = Integer.MAX_VALUE)
  int clearTeammates();

  /**
   * @since 0.1.0
   */
  boolean injure();
}

File: src/main/kotlin/me/kvdpxne/dtm/game/LocalTeam.kt
package me.kvdpxne.dtm.game

import me.kvdpxne.dtm.shared.ancillary.Communicative
import me.kvdpxne.dtm.user.LocalUser

interface LocalTeam : Communicative, Team {

  /**
   * Collection of teammates belonging to the team.
   *
   * @since 0.1.0
   */
  val teammates: List<Teammate>

  /**
   * @since 0.1.0
   */
  var health: Int

  /**
   * Returns an unsigned integer representing the total number of teammates
   * on the current team.
   *
   * @return The number of teammates on the team.
   * @since 0.1
   */
  val size: Int

  /**
   * @since 0.1.0
   */
  fun hasTeammate(
    user: LocalUser
  ): Boolean

  /**
   * @since 0.1.0
   */
  fun getTeammate(
    user: LocalUser
  ): Teammate?

  /**
   * This method is transient and should not be used directly on this object.
   * If you need to add a teammate to a team, use the [Game.addTeammate]
   * method.
   */
  fun addTeammate(
    teammate: Teammate
  ): Boolean

  /**
   * This method is transient and should not be used directly on this object.
   * If you need to remove a teammate from a team, use the
   * [Game.removeTeammate] method.
   */
  fun removeTeammate(
    user: LocalUser
  ): Boolean

  /**
   * This method is transient and should not be used directly on this object.
   * If you need to remove a teammate from a team, use the
   * [Game.removeTeammate] method.
   */
  fun removeTeammate(
    teammate: Teammate
  ): Boolean

  /**
   * @since 0.1.0
   */
  fun removeTeammates()

  /**
   * @since 0.1.0
   */
  fun injure(): Boolean
}